### PR TITLE
Add typing

### PR DESCRIPTION
**Pull request by mxr**
Feel free to close if this isn't helpful or the "tuple with optional args" implementation is not desired

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -13,6 +13,8 @@ repos:
     rev: 3.7.1
     hooks:
     -   id: flake8
+        additional_dependencies:
```
Comment (asottile) [inline .pre-commit-config.yaml]: I prefer `additional_dependencies: [flake8-typing-imports]`

File: setup.cfg [hunk 2]
```diff
@@ -23,6 +23,7 @@ classifiers =
 
 [options]
 py_modules = tokenize_rt
+install_requires = typing; python_version=='2.7'
```
Comment (asottile) [inline setup.cfg]: I think this is supposed to be `python_version="2.7"`

File: tokenize_rt.py [hunk 3]
```diff
@@ -6,35 +6,65 @@
...
+from typing import Iterable
+from typing import Iterator
+from typing import List
+from typing import Optional
+from typing import Pattern
+from typing import Sequence
+from typing import Tuple
+
+if False:
+    from typing import Text
```
Comment (asottile) [inline tokenize_rt.py]: we're not testing python2 mypy, so you can just use `str`

File: tokenize_rt.py [hunk 4]
```diff
@@ -6,35 +6,65 @@
...
-Offset.__new__.__defaults__ = (None, None)
-Token = collections.namedtuple(
+
+
+class Offset(collections.namedtuple('Offset', ('line', 'utf8_byte_offset'))):
+    __slots__ = ()
+
+    def __new__(cls, line=None, utf8_byte_offset=None):
+        # type: (Optional[int], Optional[int]) -> None
+        return super(Offset, cls).__new__(cls, line, utf8_byte_offset)
```
Comment (asottile) [inline tokenize_rt.py]: bleh, this is kinda the worst part :/

File: tokenize_rt.py [hunk 5]
```diff
@@ -6,35 +6,65 @@
...
         return s[:match.start()], s[slice(*match.span())], s[match.end():]
     else:
         return (s, '', '')
 
 
-def src_to_tokens(src):
+def src_to_tokens(src):  # type: (Text) -> List[Token]
     tokenize_target = io.StringIO(src)
-    lines = (None,) + tuple(tokenize_target)
+    lines = ('',) + tuple(tokenize_target)
```
Comment (asottile) [inline tokenize_rt.py]: hmmmm so this was actually intentional so a programming error would trigger a `TypeError` -- but I guess this is fine too 🤷‍♂

File: tokenize_rt.py [hunk 6]
```diff
@@ -82,29 +112,31 @@ def src_to_tokens(src):
...
         yield i, tokens[i]
 
 
-def parse_string_literal(src):
+def parse_string_literal(src):  # type: (Text) -> Tuple[Text, Text]
     """parse a string literal's source into (prefix, string)"""
     match = _string_re.match(src)
-    return match.group(1), match.group(2)
+    # ignore the no-match case to preserve pre-typed behavior
+    return match.group(1), match.group(2)  # type: ignore
```
Comment (asottile) [inline tokenize_rt.py]: instead of `# type: ignore`, `assert match is not None` instead

File: tokenize_rt.py [hunk 7]
```diff
@@ -82,29 +112,31 @@ def src_to_tokens(src):
     return tokens
 
 
-def tokens_to_src(tokens):
+def tokens_to_src(tokens):  # type: (Iterable[Token]) -> Text
     return ''.join(tok.src for tok in tokens)
 
 
 def reversed_enumerate(tokens):
+    # type: (Sequence[Token]) -> Iterator[Tuple[int, Token]]
```
Comment (asottile) [inline tokenize_rt.py]: this type annotation should be `Generator[Tuple[int, Token], None, None]`

Comment (mxr) [reply tokenize_rt.py]: I remember we talked about this but I don't remember what the reason is. Can you explain again? FWIW the docs mention that this syntax is OK:

https://github.com/python/mypy/blob/v0.710/docs/source/kinds_of_types.rst#Generators

File: tokenize_rt.py (hunk 6 above)
Comment (mxr) [reply tokenize_rt.py]: I thought about that but that's going to be a breaking change (`AttributeError` -> `AssertionError`). Is that OK?

File: tokenize_rt.py (hunk 4 above)
Comment (mxr) [reply tokenize_rt.py]: Yea happy to hold off until `defaults=` is backported from 3.7 and mypy understands it :/

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mxr) [reply .pre-commit-config.yaml]: Do you still want this in here even if I'm not using the `if False` check?

File: setup.cfg (hunk 2 above)
Comment (mxr) [reply setup.cfg]: Yea saw the CI error after the fact 🤦‍♂

File: tokenize_rt.py (hunk 3 above)
Comment (mxr) [reply tokenize_rt.py]: Hmm I did this to have consistency with `from __future__ import unicode_literals` (so any string literal will not be the `str` type) but I guess I can remove it. Do you still want the `flake8-typing-imports` dependency if I'm not taking advantage of it anymore here?

File: tokenize_rt.py (hunk 4 above)
Comment (mxr) [reply tokenize_rt.py]: (or something else)

File: tokenize_rt.py (hunk 5 above)
Comment (mxr) [reply tokenize_rt.py]: I see. I thought it was so you didn't have to subtract 1 in the indexing strategy below. Typing becomes annoying without this change 🤷‍♂

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mxr) [reply .pre-commit-config.yaml]: Seems that Travis can't install it for non-py3.6 environments :/

File: setup.cfg (hunk 2 above)
Comment (mxr) [reply setup.cfg]: I think this was right but I needed a newline?

File: tokenize_rt.py (hunk 5 above)
Comment (asottile) [reply tokenize_rt.py]: it's also for that, but if I made a mistake I'd get a good error

if I wanted to keep that I could

```python
class NoZeroOnlyPositive:
    def __init__(self, seq):
        self.seq = seq
    def __getitem__(self, ind):
        if ind == 0:
            raise TypeError('no!')
       else:
           return self.seq[ind]
```

File: tokenize_rt.py (hunk 3 above)
Comment (asottile) [reply tokenize_rt.py]: yeah, it should still be there in case it regresses

File: .pre-commit-config.yaml (hunk 1 above)
Comment (asottile) [reply .pre-commit-config.yaml]: ah right dang, I'll migrate this to azure pipelines then

File: tokenize_rt.py (hunk 7 above)
Comment (mxr) [reply tokenize_rt.py]: 🏓

File: tokenize_rt.py (hunk 6 above)
Comment (mxr) [reply tokenize_rt.py]: 📟

Comment (asottile) [reply tokenize_rt.py]: yes

**Comment by asottile** [comment]
a rebae should fix the tests

**Comment by mxr** [comment]
o7

**Comment by mxr** [comment]
Will look tomorrow

File: tokenize_rt.py (hunk 4 above)
Comment (mxr) [reply tokenize_rt.py]: Any ideas for tests to make sure this didn't regress?

File: tokenize_rt.py (hunk 7 above)
Comment (mxr) [reply tokenize_rt.py]: Friendly ping :D

File: tokenize_rt.py (hunk 4 above)
Comment (asottile) [reply tokenize_rt.py]: this is fine I guess 🤷‍♂

File: tokenize_rt.py (hunk 7 above)
Comment (asottile) [reply tokenize_rt.py]: oh, I explained this in the `pyupgrade` PR let me grab the copy from there:

> it is true that Generators are Iterators, but there are situations where an iterator cannot be used in place of a generator (for instance `@contextlib.contextmanager` or using `.send()`, etc.)

**Pull request merged**